cognite/client/data_classes/data_modeling/instances.py

```diff
@@ -48,6 +49,11 @@
 PropertyIdentifier = str
 
 
+class NodeOrEdgeDataDict(TypedDict):
+    properties: dict
+    source: dict
```

Comment:
`TypedDict` is used to to type hint dictionaries having a fixed schema. To quote from https://mypy.readthedocs.io/en/stable/typed_dict.html:

> a TypedDict does not define a new runtime type

Is it not helpful for users to know that the ditionary returned by `dump` has a fixed schema and know what the schema is?

Reply:
Not if the dictionary is always just written to file or send over an API. For example: 
```
from pathlib import Path
import json

my_node_data = NodeOrEdgeData(...)
Path("dumped_node_data.json).write_text(json.dumps(my_node_data.dump())
```
In the example above having a `TypeDict` does not add any readability or extra. 

I see two reasons why we should not add this extra change
1. Have one preferred way working with `NodeOrEdgeData` which is to use the `dataclass`. I see no reason to introduce flexibility just for flexibility's sake. It will just lead different developers to work with the same object using different representations.
2. If we extrapolate that all `.dump()` methods for `CogniteResources` should return an `TypedDict` that is a lot of extra code to maintain. If this does not have a solid use case for why we should do it, I do not think we should add it to the code base.

Reply:
That is to say, I am will easily change my mind on this if there is a good use case for why we should introduce such a change. But I think it can't be 'just in case' or 'it might be useful'.

Reply:
1. Why would having a more precise type hint on `dump` change the way people use `NodeOrEdgeData`?
2. I guess the more general question is how precisely we should type hint the SDK, e.g. should we use `Literal` or `str`, or `TypedDict` or `dict` when the former is applicable. I would argue that for methods that are meant to be used by external people, many of whom would need to read documentation to understand how this work, having precise type hints is a great way to improve documentation. Having more precise type hints obviously leads to more code to maintain.

Just my two cents.